mediatek: filogic: use nvmem for Unifi 6 Plus WiFi mac

bmork opened the pull request:
Avoids the occasional lost race, where VAPs are created before the hotplug script has updated the phy mac address.

Run tested

bmork committed:
mediatek: filogic: use nvmem for Unifi 6 Plus WiFi mac

Avoids the occacial lost race, where VAPs are created before the hotplug
script has updated the phy mac address.

Signed-off-by: Bjørn Mork <bjorn@mork.no>
Link: https://github.com/openwrt/openwrt/pull/20154
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

bmork commented on target/linux/mediatek/dts/mt7981a-ubnt-unifi-6-plus.dts:
```diff
@@ -169,5 +171,19 @@
+
+	band@0 {
+		reg = <0>;
+		nvmem-cells = <&macaddr_eeprom_6 0>;
```
IIUC the eeprom contains only generic common mac addresses here, so relying on that won't do.

I don't know what the status is wrt nvmem support for emmc which is why I left that out for now. It's not as much of a problem either.

castiel652 replied:
@neheb default behaviour reads from `0x4`.

castiel652 replied:
@bmork based on this https://github.com/openwrt/openwrt/commit/dbc2923cbe21ad9ea3fbdddc4b880a6e81678e11 I think it's supported.

neheb replied:
MAC starting at 6 is interesting…

bmork replied:
Thanks for the pointer!  I've added another commit for the caldata now.  Working like a charm:

Before

r```
oot@u6plus:~# sha256sum  /sys/kernel/debug/ieee80211/phy*/mt76/eeprom  /lib/firmware/mediatek/mt7981_eeprom_mt7976_dbdc.bin
3a81e398480c05230c968bbdd3808e9fa33328c61a1399b4c933023111c5e69a  /sys/kernel/debug/ieee80211/phy0/mt76/eeprom
3a81e398480c05230c968bbdd3808e9fa33328c61a1399b4c933023111c5e69a  /sys/kernel/debug/ieee80211/phy1/mt76/eeprom
3a81e398480c05230c968bbdd3808e9fa33328c61a1399b4c933023111c5e69a  /lib/firmware/mediatek/mt7981_eeprom_mt7976_dbdc.bin

```

After:

```
root@u6plus:~# sha256sum  /sys/kernel/debug/ieee80211/phy*/mt76/eeprom  /lib/firmware/mediatek/mt7981_eeprom_mt7976_dbdc.bin
3a81e398480c05230c968bbdd3808e9fa33328c61a1399b4c933023111c5e69a  /sys/kernel/debug/ieee80211/phy0/mt76/eeprom
3a81e398480c05230c968bbdd3808e9fa33328c61a1399b4c933023111c5e69a  /sys/kernel/debug/ieee80211/phy1/mt76/eeprom
sha256sum: can't open '/lib/firmware/mediatek/mt7981_eeprom_mt7976_dbdc.bin': No such file or directory

```

castiel652 replied:
Do you have `missing precal` in bootlog

bmork replied:
The two mac addresses at the beginning of the nor EEPROM partition are actually the same with and without the local bit:

```
root@u6plus:~# hexdump -n 12 -C /dev/mtd0
00000000  e4 38 83 e5 85 d4 e6 38  83 e5 85 d4              |.8.....8....|
0000000c
```

We've always used the second one for the wifi phys.

As for the offsets:  This is NOT an mt76 eeprom despite the confusing partition name.  It's a collection of system variables which could have been stored in an eeprom if there was one. The mt76 eeprom caldata is stored in the "factory" emmc partition and is not available from the nor flash.  And the mac addresses are not stored anywhere in the emmc I believe

bmork replied:
> Do you have `missing precal` in bootlog

Yes, I do.  What does that mean and is it something which should/can be fixed?

The message has always been there.  Used to say

`[   15.150723] mt798x-wmac 18000000.wifi: missing precal data, size=200720`

but recently changed the reported "size" to

`[   15.558514] mt798x-wmac 18000000.wifi: missing precal data, size=454672`

castiel652 replied:
There is regular calibration when there's no precal data.
And from the commit message
```
Note that group_cal can save 30% bootup calibration time, and dpd_cal can
save 75% channel switching time.
```

bmork committed:
mediatek: filogic: use nvmem for Unifi 6 Plus caldata

Replace the hotplug script with nvmem now that loading data from emmc
partitions is supported.

Signed-off-by: Bjørn Mork <bjorn@mork.no>
Link: https://github.com/openwrt/openwrt/pull/20154
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

castiel652 commented on target/linux/mediatek/dts/mt7981a-ubnt-unifi-6-plus.dts:
```diff
@@ -169,5 +198,21 @@
 };
 
 &wifi {
+	nvmem-cells = <&eeprom_factory_0>;
+	nvmem-cell-names = "eeprom";
```
```suggestion
	nvmem-cells = <&eeprom_factory_0>, <&precal_factory_1010>;
	nvmem-cell-names = "eeprom", "precal";
```

bmork committed:
mediatek: filogic: load precal data for Unifi 6 Plus

Point the driver to the precal section of the emmc "factory" partition.

Signed-off-by: Bjørn Mork <bjorn@mork.no>
Link: https://github.com/openwrt/openwrt/pull/20154
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
